### DISCUSSION TITLE
The Gecko Zigbee Python interface project

### DISCUSSION THREAD

**Discussion by MattWestb** [zigpy/zigpy#598] [Ideas]
The ZbPy [Gecko Zigbee Python interface](https://github.com/osresearch/ZbPy) project have implementing the hardware and the radio part of the EFR32 and is having the  IEEE802.15.4 and ZigBee network stack up and running in micro python.
They can joining and decrypting zigbee messages but have only partly implanting the ZCL for cluster mapping.

I finding the project interesting but i dont know is the first gen EFR32 have enough hardware resources for running both the hardware, radio and Zigbee cluster labory at the same time without problems. Perhaps with second gen its possible running one stripped Zigpy that can have one limited basic quirks and sending / transmitting "clean" ZCL commands on the serial port to / from the host system.

Can it being one future for ZHA having more EFR32 running "Micro Zigpy" and talking with ZHA thru ser2net (or some other good working protocol that is suit for our purposes) and ZHA that is knowing where the devices is resident and linking the funktionality to respective "zigbee concentrator" ?

Then only using ZLL or ZHA devices its possible striping the device suport down and using multiple concentrators in large buildings and only need one central ZHA instance for doing the higher level operations and still getting one good scalable and flexible system.
Zigbee 3 can't growing so large as one old pre ZB3 ZLL / ZHA system was able to do (key handling and other security stuff is limiting it).
Also making one "mega large Zigbee network" is getting problem if it being to dense (radio layer) and the latency is also being one large problem then messages must being relayed thru one large mesh perhaps with around 100 or more hopes for one message from source to destination.

Good and bad thing: Only one processor of doing the underlayer zigbee thing (more CPUs / system = more problems) and one upper zigbee layer (ZHA) making it scalable and clean.
Its not redundant for communication problems between concentrator and ZHA and its one complex custom component that need software mentrance also in the under layer for keeping it up to date.
Is it better using one more hard/software layer with one ESP32 (or somthing else suitable) that is hosting the middle zigbee part (ESP-zigpy?) and have more intelligence between the under zigbee layer and the upper zigbee layer (ZHA) and can doing automations and other thing there ?

Only one more crazy brainstorming from my side :-))

**Comment by MattWestb** [comment]
I think [Z-Smart Systems](https://github.com/zsmartsystems/com.zsmartsystems.zigbee) is having multi ZNCP up and running but i have not digging deeper in there system docks.
And there Zigbee sniffer is working with EZSP dongles :-)))

**Comment by Hedda** [comment]
@osresearch What did you have in mind when you wrote this comment in your [ZbPy](https://github.com/osresearch/ZbPy) project about porting parts of zigpy to MicroPython?

"_The Application and Cluster layers are only partially implemented right now; the messages are decoded but not mapped to functionality. Perhaps parts of ZigPy can be ported to MicroPython to provide these mappings._"

https://github.com/osresearch/ZbPy

https://github.com/osresearch/micropython/tree/efm32/ports/efm32

https://trmm.net/Ikea/

PS: I guess that you are probably already aware of MattWestb's IKEA Trådfri hardware docs/projects too? https://github.com/MattWestb/IKEA-TRADFRI-ICC-A-1-Module and that some new devices in the IKEA Tradfri series come a newer EFR32MG21 (EFR32 Mighty Gecko Series 2) SoC based module https://github.com/MattWestb/IKEA-TRADFRI-ICC-A-1-Module#new-ikea-tr%C3%A5dfri-markus-ezsp-module

**Comment by osresearch** [comment]
I've been away from the ZbPy project for a year, so I have to dig back into my memory for what I meant... I also haven't bought any new bulbs in a while, so I hadn't seen the new modules or been keeping up with other people's hacking.  More SRAM will make micropython much happier, so I'm glad to see that they have moved to a 96KB SoC.

I think my thought was that building my own ZCL module was duplicating effort, so if I could figure out a way to have a clean handoff from the EFR32 micropython to ZbPy to ZigPy, then it would be a better way to build the device or coordinator functionality.  At the time my micropython didn't support async await, so it wasn't possible to use ZigPy directly, although I think that is now supported.

The roadblock that I ran into (back in early 2020) was passing the coordinator interview and I ran out of steam debugging it.  If I had the EFR32 already joined to the network, then the bare-bones ZCL could handle on-off commands and brightness, but for some reason zigbee2mqtt never approved of my device during the interview (and the debugging / logging was so minimal that I couldn't figure out what it was expecting).  If that is fixed, then the ZbPy library is almost ready for some limited applications (and if ZigPy is used instead, then perhaps even wider ones)

**Reply by Hedda** [reply]
> haven't bought any new bulbs in a while, so I hadn't seen the new modules or been keeping up with other people's hacking. More SRAM will make micropython much happier, so I'm glad to see that they have moved to a 96KB SoC.

Yes the new [new IKEA Trådfri GU10 RGB/RGBW LED1923R5 (Article Number 804.392.28 in Europe) with the newer "MGM210LA22JNF2" / "MGM210L" module which MCU has a CPU with up to 80 MHz core frequency (double that of EFR32MG1 Series 1), 1024 kB flash and 96 kB RAM + a 20 dBm radio but as I understand are so far still rare to find and not yet widely available in all IKEA stores. More info in https://github.com/basilfx/TRADFRI-Hacking/issues/39 now. I hope and would expect, however, that more new products in the IKEA Tradfri series will probably be based on that newer module in the future. 

Though for development purposes you might also be interested in https://github.com/zigpy/zigpy/discussions/586 as well simply because of their current availability.

Also, ITead will soon be releasing an inexpensive Sonoff branded EFR32MG21 based Zigbee USB dongle (more info in https://github.com/zigpy/zigpy/discussions/635 ).

**Comment by mguaylam** [comment]
Might be out of context but I would love to replace the firmware of the IKEA devices with something less buggy and much more flexible like more dimming and power on state..

**Reply by MattWestb** [reply]
If you is buying one dev kit (**that is qualified for zigbee access, all is not**) you can cooking your own firmware from standard examples and modifying them as you like.

The big problems is loading it on the devices then OTA is not possible then you cant signing the firmware with IKEAs private key so first need flashing one new bootloader with SWD or J-tag.
And getting access to the inside of some products is not so easy without destroying them (its one security thing with high voltage devices) and other is very easy.

Its little rear but some of the users in the community is cooking there own hard and / or software and its working but needs experience for getting it good.

For flashing bootloader and dumping firmware from IKEA devices i have documented in my git also how to using the new bootloader with standard cooked firmware.
![E1743](https://user-images.githubusercontent.com/49618193/106249776-13249f80-6213-11eb-9da6-5f34377c5c72.jpg)
One E1743 easely out of the case and not destroyed hooked up well dumped both firmware and user data then new bootloader and EZSP firmware so can being user as coordinator but is normally my Zigbee sniffer device :-)))

**Reply by mguaylam** [reply]
Nice! What do you use to capture its frames between the Silabs and Wireshark?

**Reply by MattWestb** [reply]
I have using the instruction from [With HUSBZB-1 stick](https://www.zigbee2mqtt.io/how_tos/how_to_sniff_zigbee_traffic.html)  = https://github.com/zsmartsystems/com.zsmartsystems.zigbee.sniffer.
